### PR TITLE
Fixes #26409 - Locking of packages after installation

### PR DESCRIPTION
**Pull request by mbacovsky**
This patch add a new option --[no-]lock-package-versions.
The option enables/disables version locking of installed packages.
Once used the choice is stored in the scenario "custom" hash and
is reused in the following installer runs.

If the packages are locked when the installer starts
unlock is performed before installer proceed with the installation.

### CODE REVIEW

**Comment by mbacovsky** [comment]
This PR requires version locking feature in foreman-maintain to be merged https://github.com/theforeman/foreman_maintain/pull/249

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]:
```suggestion
    Kafo::Helpers.log_and_say :info, "Package versions are locked. Continuing with unlock."
```

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]: Exit status 1 surprises me here. I would expect that you could do `foreman-maintain installation is-locked --assumeyes && foreman-maintain installation unlock --assumeyes` on the shell to ensure it's unlocked. Similarly `foreman-maintain installation is-locked --assumeyes || foreman-maintain installation lock --assumeyes` to lock.

**Review by ekohl — COMMENTED** [review]
Untested, but overall it looks pretty clean.

**Comment by mbacovsky** [comment]
@ekohl thanks for the review. I fixed the typo and changed the exitcode in formean-maintain and updated here. The changes are in separate commit and needs to be squashed on merge.

**Comment by mbacovsky** [comment]
Updated, the `foreman-maintain installation` was renamed to `foreman-maintain packages`.

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]: If this failed and the user indicated they want to lock this should be fatal.

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]: I assume unlocking is idempotent so do we even need to check if it's locked? IIRC the execute helper also prints + logs the output so the log line might not be needed. This will mostly depend on the output of foreman-maintain and I'll leave this up to you.

File: katello/hooks/boot/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/boot/09-version_locking.rb]: One thought to implement this in regular foreman is to check if foreman-maintain is present and only then add the option. Probably best to leave that out of this PR though since it introduces complexities.

**Review by ekohl — COMMENTED** [review]
Edit: started to write something here and then applied to lines.

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (mbacovsky) [reply katello/hooks/pre_commit/09-version_locking.rb]: @ekohl the unlock will probably be expensive on Debian so I wanted to do it only when necessary. On yum based systems it is probably not the case and the check could be skipped. I'll keep the foreman-maintain output in log only.

File: katello/hooks/boot/09-version_locking.rb
Comment (mbacovsky) [reply katello/hooks/boot/09-version_locking.rb]: I made an attempt to handle missing foreman-maintain. Though I guess it is a corner case as this is katello-only and we depend on f-m there.

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (mbacovsky) [reply katello/hooks/pre_commit/09-version_locking.rb]: Added some more checks

**Comment by mbacovsky** [comment]
@ekohl I've added some more error handling and tuned the output. Please check. It is in a new commit.

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]: Nit: empty line

File: katello/hooks/pre_commit/09-version_locking.rb
Comment (ekohl) [inline katello/hooks/pre_commit/09-version_locking.rb]: This still makes it a hard requirement. Should this be inside `if lock_versions`?

Comment (mbacovsky) [reply katello/hooks/pre_commit/09-version_locking.rb]: @ekohl this is outside `lock_versions` intentionally for the case the packages were locked manually and auto-locking is not enabled in installer. What behavior of installer would you expect in such case?

It is not actually hard req as the condition is negative also when the command does not exist. We may add `2>&1` to hide the error out though.

Comment (ekohl) [reply katello/hooks/pre_commit/09-version_locking.rb]: You're right, I totally missed that. Nice :)

Comment (mbacovsky) [reply katello/hooks/pre_commit/09-version_locking.rb]: @ekohl I assume it is okay to let installer unlock manually locked packages, please let me know if it is not. I also suppressed the error out for clean output. Updated.

Comment (ekohl) [reply katello/hooks/pre_commit/09-version_locking.rb]: IMHO as long as it's documented, that's fine. Users can still opt of out locking and handle it manually if I read the code correct. That's what we would advise users.

**Comment by mbacovsky** [comment]
@ekohl now when the PR does not depend on the foreman-maintain what prevents it from being merged? Is there anything I can help with?

**Comment by jameerpathan111** [comment]
@mbacovsky @ekohl 
I have tested this PR, it LGTM, and it works as expected.
I haven't found any major issues in it yet.

```
# ./bin/foreman-maintain packages is-locked
Foreman related packages are not locked
```
```
# yum update
=========================================================================================================================================================================================
 Package                                                 Arch         Version                        Repository                                                                     Size
=========================================================================================================================================================================================
Installing:
 satellite-common                                        noarch       6.5.0-11.el7sat                Sat6-CI_Red_Hat_Satellite_6_5_Composes_Satellite_6_5_RHEL7                     15 k
     replacing  rh-ruby24-ruby.x86_64 2.4.5-91.el7
...
...#skipping full list of packages being updated
Updating:
 ansiblerole-insights-client                             noarch       1.6-1.el7sat                   Sat6-CI_Red_Hat_Satellite_6_5_Composes_Satellite_6_5_RHEL7                     17 k
...
...#skipping full list of packages being updated
Transaction Summary
=========================================================================================================================================================================================
Install    1 Package  (+77 Dependent packages)
Upgrade  220 Packages

Total download size: 472 M
Is this ok [y/d/N]: n

```
```
# satellite-installer --lock-package-versions
Resetting puppet server version param...
Installing             Done                                               [100%] [......................................................................................................]
Package versions are being locked.
```
```
# yum update

=========================================================================================================================================================================================
 Package                              Arch                                    Version                                          Repository                                           Size
=========================================================================================================================================================================================
Updating:
 zsh                                  x86_64                                  5.0.2-31.el7                                     rhel-7-server-rpms                                  2.4 M

Transaction Summary
=========================================================================================================================================================================================
Upgrade  1 Package

Total download size: 2.4 M
Is this ok [y/d/N]: n

```
```
# ./bin/foreman-maintain packages is-locked
Foreman related packages are locked
```
```
# satellite-installer --no-lock-package-versions
Resetting puppet server version param...
Package versions are locked. Continuing with unlock.
Installing             Done                                               [100%] [......................................................................................................]
  Success!
```

**Review by ekohl — APPROVED** [review]
I don't see any more objections, but I'm holding off on merging till I've completed all the module releases. Will look over this before branching.

**Comment by ekohl** [comment]
Looks like I forgot to squash on merge.

**Pull request merged**
